compendium/DeclarativeServices/src/SCRActivator.cpp

```diff
@@ -143,30 +144,21 @@ namespace cppmicroservices
                 return;
             }
 
-            bool extensionFound = false;
-            {
-                std::lock_guard<std::mutex> l(bundleRegMutex);
-                extensionFound = (bundleRegistry.count(bundle.GetBundleId()) != 0u);
-            }
-
-            // bundle components have not been loaded, so create the extension which will load the components
-            if (!extensionFound)
+            //create the extension which will load the components
+            if (!bundleRegistry->Find(bundle.GetBundleId()))
             {
                 logger->Log(SeverityLevel::LOG_DEBUG, "Creating SCRBundleExtension ... " + bundle.GetSymbolicName());
                 try
                 {
                     auto const& scrMap = ref_any_cast<cppmicroservices::AnyMap>(headers.at(SERVICE_COMPONENT));
-                    auto ba = std::make_unique<SCRBundleExtension>(bundle,
-                                                                   scrMap,
+                    auto ba = std::make_shared<SCRBundleExtension>(bundle,
```

Comment:
If this is a unique_ptr then I would have to use std::move in the bundleRegistry->Add call and it wouldn't be available for the ba->Initialize call. 

The reason I had to split the functionality from the old SCRBundleExtension constructor into a constructor and an Initialize method is because factory components can be added to ConfigurationAdmin before a bundle is started. When the bundle is started, DS will find the configuration objects and call CreateFactoryComponent to create them. For CreateFactoryComponent to work it has be able to find the SCRBundleExtension object in the SCRExtensionRegistry 
 and then call the SCRBundleExtension::AddComponentManager method to add the ComponentManagerImpl objects to the SCRBundleExtension::managers map. 

So the new constructor creates the managers map. SCRActivator adds the SCRBundleExtension object to the SCRExtensionRegistry. At this point the Initialize method can be called to create the ComponentConfiguration objects and if CreateFactoryComponent is called it will work fine.